damnit/api.py

```diff
@@ -0,0 +1,250 @@
+        with h5py.File(self._h5_path) as f:
+            yield f[self.name]
+
+    @staticmethod
+    def _type_hint(group):
+        hint_s = group.attrs.get('_damnit_objtype', '')
+        if hint_s:
+            return DataType(hint_s)
+        return None
+
+    def _read_netcdf(self, one_array=False):
+        load = xr.load_dataarray if one_array else xr.load_dataset
+        obj = load(self._h5_path, group=self.name, engine="h5netcdf")
+        # Remove internal attributes from loaded object
+        obj.attrs = {k: v for (k, v) in obj.attrs.items()
+                     if not k.startswith('_damnit_')}
+        return obj
+
+    def read(self):
+        if self._db_only:
+            return self.summary()
+
+        with self._open_h5_group() as group:
+            type_hint = self._type_hint(group)
+            if type_hint is DataType.Dataset:
+                return self._read_netcdf()
+            elif type_hint is DataType.DataArray:
+                return self._read_netcdf(one_array=True)
+
+            dset = group["data"]
+            if h5py.check_string_dtype(dset.dtype) is not None:
+                # Strings. Scalar/non-scalar strings need to be read differently.
+                if dset.ndim == 0:
+                    return dset[()].decode("utf-8", "surrogateescape")
+                else:
+                    return dset.asstr("utf-8", "surrogateescape")[0]
+            elif dset.ndim == 0:
+                # Scalars
+                return dset[()]
+            else:
+                # Otherwise, return a Numpy array
+                return group["data"][()]
+
+    def summary(self):
+        result = self._db.conn.execute("""
+            SELECT value, max(version) FROM run_variables
+            WHERE proposal=? AND run=? AND name=?
+        """, (self.proposal, self.run, self._name)).fetchone()
+
+        if result is None:
+            # This should never be reached unless the variable is deleted
+            # after creating the VariableData object.
+            raise RuntimeError(f"Could not find value for '{self._name}' in p{self.proposal}, r{self._name}")
+        else:
+            return result[0]
+
+    def __repr__(self):
+        return f"<VariableData for '{self.name}' in p{self.proposal}, r{self.run}>"
+
+
+class RunVariables:
+    def __init__(self, db_dir, run):
+        self._db = DamnitDB.from_dir(db_dir)
+        self._proposal = self._db.metameta["proposal"]
+        self._run = run
+        self._data_format_version = self._db.metameta["data_format_version"]
+        self._h5_path = Path(db_dir) / f"extracted_data/p{self._proposal}_r{self._run}.h5"
+
+    @property
+    def proposal(self):
+        return self._proposal
+
+    @property
+    def run(self):
+        return self._run
+
+    def __getitem__(self, name):
+        key_locs = self._key_locations()
+        if name not in key_locs:
+            raise KeyError(f"Variable data for '{name}' not found for p{self.proposal}, r{self.run}")
```

Comment:
```suggestion
            raise KeyError(f"Variable data for {name!r} not found for p{self.proposal}, r{self.run}")
```

This quotes the string and takes care of escaping any quotes inside it, so you can tell from the error message exactly what string got here, which could be more important now that we're allowing titles.

Reply:
Fixed in d9e9f9b.